File: etl/dags/tools/opa.py
import json
import logging
import os.path
from typing import Any

from airflow.exceptions import AirflowFailException
from airflow.hooks.subprocess import SubprocessHook, SubprocessResult
from airflow.models import Variable

from dags.tools import fs

logger = logging.getLogger("airflow.task")


def run(
    *,
    build_script_path: str,
    protected_branch: str,
    repository_name: str,
    repo_url: str,
    provenance_url: str | None,
    token: str,
) -> dict[str, Any]:
    scribe_service_bin_path = Variable.get("scribe_service_bin_path")

    policy_path = Variable.get("github_posture_policy_path")
    policy_path = os.path.join(policy_path, "github")

    out_file = fs.get_tmp_file_path()

    result = _call_subprocess(
        scribe_service_bin_path,
        out_file,
        policy_path,
        build_script_path,
        protected_branch,
        repository_name,
        repo_url,
        token,
        provenance_url or "",
    )
    if result.exit_code != 0:
        logger.error("opa failed with exit_code: %s", result.exit_code)
        raise AirflowFailException("opa failed")

    with open(out_file) as f:
        sarif_data = json.load(f)
    fs.remove_file(out_file)
    return sarif_data


def _call_subprocess(
    scribe_service_bin_path: str,
    out_file: str,
    policy_path: str,
    build_script_path: str,
    protected_branch: str,
    repository_name: str,
    repo_url: str,
    token: str,
    provenance_url: str,
) -> SubprocessResult:
    """
    separated function for testing purposes
    """
    return SubprocessHook().run_command(
        [
            scribe_service_bin_path,
            "opa",
            "--build-script-path",
            build_script_path,
            "--out",
            out_file,
            "--policy-path",
            policy_path,
            "--protected-branch",
            protected_branch,
            "--provenance-file-url",
            provenance_url,
            "--repository-name",
            repository_name,
            "--repository-url",
            repo_url,
            "--token",
            token,
        ]
    )

File: scribe-service/cmd/main/opa.go
package main

import (
	"log"

	"github.com/spf13/cobra"

	"github.com/scribe-security/scribe2/scribe-service/internal/rego"
)

var (
	provenanceFileURL string
	protectedBranch   string
	token             string
	policyPath        string
	buildScriptPath   string
	repositoryName    string
	repositoryURL     string
	debug             bool
	// provenanceURL     string
	outFile string
)

// sigVerifyCmd represents the sigVerify command
var opaCmd = &cobra.Command{
	Use: "opa",
	Run: func(cmd *cobra.Command, args []string) {
		err := rego.InvokePolicy(
			cmd.Context(),
			provenanceFileURL,
			protectedBranch,
			token,
			policyPath,
			buildScriptPath,
			repositoryName,
			repositoryURL,
			debug,
			outFile)
		if err != nil {
			log.Fatal(err)
		}
	},
}

func init() {
	opaCmd.Flags().StringVarP(&policyPath, "policy-path", "p", "", "Path to the policy")
	must(opaCmd.MarkFlagRequired("policy-path"))

	// provenanceFileURL is not required
	opaCmd.Flags().StringVarP(&provenanceFileURL, "provenance-file-url", "f", "", "Provenance file URL")

	opaCmd.Flags().StringVarP(&protectedBranch, "protected-branch", "b", "", "Protected branch")
	must(opaCmd.MarkFlagRequired("protected-branch"))

	opaCmd.Flags().StringVarP(&token, "token", "t", "", "Token")
	must(opaCmd.MarkFlagRequired("token"))

	opaCmd.Flags().StringVarP(&buildScriptPath, "build-script-path", "s", "", "Build script path")
	must(opaCmd.MarkFlagRequired("build-script-path"))

	opaCmd.Flags().StringVarP(&repositoryName, "repository-name", "r", "", "Repository name")
	must(opaCmd.MarkFlagRequired("repository-name"))

	opaCmd.Flags().StringVarP(&repositoryURL, "repository-url", "u", "", "Repository URL")
	must(opaCmd.MarkFlagRequired("repository-url"))

	opaCmd.Flags().StringVarP(&outFile, "out", "o", "", "Out")
	must(opaCmd.MarkFlagRequired("out"))

	opaCmd.Flags().BoolVarP(&debug, "debug", "d", false, "Debug")

	rootCmd.AddCommand(opaCmd)
}

func must(err error) {
	if err != nil {
		log.Fatal(err)
	}
}
